### PR TITLE
new: logout: avoid an open redirect

### PR DESCRIPTION
**Pull request by vladimir-mencl-eresearch**
Only accept a return URL passed to /logout if it is either:
* an HTTPS URL pointing to the same host
* a local path relative URL

Validate the URL through Ruby URI.parse as part of the process.

Log rejected URLs for audit purposes.

Note: keeping the return URL parameter for logout is useful for supporting a logout hook for Shibboleth SP - and the local URL restriction will work with it well.

https://wiki.shibboleth.net/confluence/display/SP3/Notify

### CODE REVIEW

**Comment by James-REANNZ** [comment]
Looks ok to me.

Do we support `localhost` cases (on TEST at least)? If so, it may be worth also allowing `localhost` + `http` cases to support people developing apps locally.

**Comment by vladimir-mencl-eresearch** [comment]
Thanks for the quick review @James-REANNZ !

I did consider that as well, but in the end decided for simple rules.  And redirecting to "HTTPS only" (no caveats) is a clear rule for security.

And I did not want to introduce special exceptions for development of RapidConnect.

If you mean it for development of applications connecting to RapidConnect ... well, as this PR stands, it does not support redirecting back to the remote app at all, only to RapidConnect itself.

There is only one redirect back to an app in use, for NeSI Zendesk.  This is something I configured years ago (while still working for NeSI) - and that is also when+why I added the return URL feature to RapidConnect's logout.

But continuing to support this would be tricky.  I could compare the return URL to the registered URL of an app - but at the moment, the logout URL does not get any reference to the app requesting the logout.  (Hm, except for possibly the `Referrer` header).

To continue supporting redirect to app, I'd have to extend the logout handler "API" to also receive an app ID - and if passed, it would also accept URLs with the same scheme and base hostname as the app.

However, that would open the door for someone to register a malicious base URL (auto-accepted in TEST) and then validly redirect to that URL.  And our TEST could be used as an open redirect - same reputational damage as in PROD.

That leads me to not supporting logout redirects outside RapidConnect.

In the end, the only thing this takes away is being able to redirect somewhere else on logout - which is unlikely to be a major pain-point.  (I'll still have to let NeSI know this will change behaviour when someone logs out of their Zendesk).

In summary, the only safe way to allow redirecting to app would be to:
* change API to also get app ID
* disable auto-approval in TEST
* work with deployments (NeSI) to change to new API

Given that RapidConnect will likely soon be replaced by the RapidConnect module on FederationManager, I lean towards not investing too much here.

Your thoughts?

Cheers,
Vlad

**Comment by James-REANNZ** [comment]
Given the low impact, it's probably fine as-is.

**Comment by vladimir-mencl-eresearch** [comment]
Thanks!
